criterion-measurement/src/Criterion/Measurement.hs

```diff
@@ -289,41 +291,69 @@ runBenchmark :: Benchmarkable
              -- exceeded in order to generate enough data to perform
              -- meaningful statistical analyses.
              -> IO (V.Vector Measured, Double)
-runBenchmark bm timeLimit = do
+runBenchmark bm timeLimit = runBenchmarkWith endAt 0 bm where
+  endAt :: Int -> Int64 -> Double -> NonEmpty Measured -> Double -> Maybe (Int64, Double)
```

Comment:
Since this is the default behavior for `runBenchmarkWith`'s termination check, I'd favor factoring this out into a top-level definition so that a user can use it if they wish to tweak it slightly. Since this definition closes over `timeLimit`, it might make things more clear to turn this into a newtype:

```hs
newtype TerminationCheck s = TerminationCheck
  { getTerminationCheck :: Int -> Int64 -> Double -> NonEmpty Measured -> s -> Maybe (Int64, s) }

defaultTerminationCheck :: Double -> TerminationCheck s
defaultTerminationCheck timeLimit = TerminationCheck $ \count iters delta ms prev -> do ...
```

Then you can mention in the Haddocks for `runBenchmark` that `runBenchmark bm timeLimit = runBenchmarkWith (defaultTerminationCheck timeLimit) 0 bm`.

Reply:
I'm not so sure about. That function needs to be factored so its pieces could be  usable, but I don't know how we could factor it. There are at least two parts: updating the state and deciding termination, but that consist of three checks already.

Reply:
I think I explained myself poorly in https://github.com/haskell/criterion/pull/255#discussion_r783494620, so let me try again. I agree that because termination checks will need to perform arbitrary computation, there's not a good way in general to configure everything that a termination check might want to do. That being said, I still think there is value in exposing the default termination check to users. This is because in #218, a user wishes to do what `runBenchmark` does plus an additional constraint on the number of iterations performed. In terms of code, this could look something like this:

```hs
longLivedComputationCheck :: TerminationCheck r s
longLivedComputationCheck = TerminationCheck $ \... -> do
  getTerminationCheck defaultTerminationCheck ...
  if <time exceeds certain threshold>
    then Left ...
    else Right ...
```

For some use cases, there might not be a better alternative than having to copy-paste the source code of `defaultTerminationCheck` and tweaking it, but I don't want that to always be the case.